### PR TITLE
Optimized Visual.IsFlat implementation to avoid subtree traversion

### PR DESCRIPTION
**Pull request by Duckers**
Don't merge yet, will need more tests

### CODE REVIEW

**Comment by Duckers** [comment]
There - we have good test coverage and should be all green

File: Source/Fuse.Nodes/Transform.uno [hunk 1]
```diff
@@ -29,8 +29,17 @@ namespace Fuse
 		
 		internal event Action<Transform> MatrixChanged;
 
+		bool _wasFlat;
```
Comment (mortoray) [inline Source/Fuse.Nodes/Transform.uno]: This doesn't appear to be initialized anywhere.

File: Source/Fuse.Nodes/Visual.IsFlat.uno [hunk 2]
```diff
@@ -8,92 +9,74 @@ namespace Fuse
...
-		
-		bool _areChildrenFlatCached
+
+		bool _flatRooted;
+		void FlatRooted()
 		{
-			get { return HasBit(FastProperty1.AreChildrenFlatCached); }
-			set { SetBit(FastProperty1.AreChildrenFlatCached, value); }
+			if (_nonFlat != 0)
+				throw new Exception(); // Should never happen
```
Comment (mortoray) [inline Source/Fuse.Nodes/Visual.IsFlat.uno]: A message would be nice.

File: Source/Fuse.Nodes/Visual.uno [hunk 3]
```diff
@@ -147,6 +148,8 @@ namespace Fuse
 				}
 			}
 
+			FlatRooted();
```
Comment (mortoray) [inline Source/Fuse.Nodes/Visual.uno]: Shouldn't this be FlatUnrooted() ?

File: Source/Fuse.Nodes/Visual.IsFlat.uno [hunk 4]
```diff
@@ -8,92 +9,74 @@ namespace Fuse
...
+		void FlatUnrooted()
 		{
-			get { return HasBit(FastProperty1.AreChildrenFlat); }
-			set { SetBit(FastProperty1.AreChildrenFlat, value); }
+			if (_localNonFlat)
+				DecrementNonFlat();
+			_flatRooted = false;
+
+			if (_nonFlat != 0)
+				throw new Exception(); // Should never happen
```
Comment (mortoray) [inline Source/Fuse.Nodes/Visual.IsFlat.uno]: Add a message to the exception.

File: Source/Fuse.Nodes/Transform.uno (hunk 1 above)
Comment (Duckers) [inline Source/Fuse.Nodes/Transform.uno]: In Uno, fields are initialized to zero/false by default.

File: Source/Fuse.Nodes/Visual.IsFlat.uno (hunk 2 above)
Comment (Duckers) [inline Source/Fuse.Nodes/Visual.IsFlat.uno]: Users are not supposed to see this, only for tests to trigger.

File: Source/Fuse.Nodes/Visual.IsFlat.uno (hunk 4 above)
Comment (Duckers) [inline Source/Fuse.Nodes/Visual.IsFlat.uno]: Users are not supposed to see this, only for tests to trigger.

File: Source/Fuse.Nodes/Transform.uno (hunk 1 above)
Comment (mortoray) [inline Source/Fuse.Nodes/Transform.uno]: Yes, I know, but a Transform may not be flat by default. Some of the constructors take arguments, don't they?

**Comment by kusma** [comment]
So, this is clearly broken ;)

> Failed:  Fuse.Controls.Test.ViewportTest.Cache
> Failed:  Fuse.Controls.Test.ViewportTest.IsFlat
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.ColorPanel
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.ColorRectangle
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.CornerRadius
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.Falloff
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.InvisiblePanel
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.InvisibleRectangle
> Failed:  Fuse.Controls.Primitives.Test.ShadowTest.Padding

**Comment by mortoray** [comment]
Replaced with https://github.com/fusetools/fuselibs-public/pull/420

**Pull request closed without merging**